test/cljfx/composite_test.clj

```diff
@@ -0,0 +1,116 @@
+                    {:fx/type lifecycle
+                     :a 2}
+                    {::foo nil})
+        _ (fact (grab-history)
+                => [{:op :replace!, :prop :a, :instance :obj, :coerced-new-value "2", :old-value 42, :new-value 2}
+                    {:op :retract!, :prop :b, :instance :obj, :coerced-value "24", :value 24}])
+        _ (fact (component/instance component)
+                => :obj)
+
+        ;; replace :a, assign :b/:c
+        component (lifecycle/advance
+                    lifecycle
+                    component
+                    {:fx/type lifecycle
+                     :a 1
+                     :b 2
+                     :c 3}
+                    nil)
+        _ (fact (grab-history)
+                => [{:op :replace!, :prop :a, :instance :obj, :coerced-new-value "1", :old-value 2, :new-value 1}
+                    {:op :assign!, :prop :b, :instance :obj, :coerced-value "2", :value 2}
+                    {:op :assign!, :prop :c, :instance :obj, :coerced-value "3", :value 3}])
+        _ (fact (component/instance component)
+                => :obj)
+
+        ;; replace :a/:b/:c
+        component (lifecycle/advance
+                    lifecycle
+                    component
+                    {:fx/type lifecycle
+                     :a 1
+                     :b 2
+                     :c 3}
+                    nil)
+        _ (fact (grab-history)
+                => [{:op :replace!, :prop :a, :instance :obj, :coerced-new-value "1", :old-value 1, :new-value 1}
+                    {:op :replace!, :prop :b, :instance :obj, :coerced-new-value "2", :old-value 2, :new-value 2}
+                    {:op :replace!, :prop :c, :instance :obj, :coerced-new-value "3", :old-value 3, :new-value 3}])
+        _ (fact (component/instance component)
+                => :obj)
+
+        ;; replace :a/:b/:c
+        component (lifecycle/advance
+                    lifecycle
+                    component
+                    {:fx/type lifecycle
+                     :a 1
+                     :b 2
+                     :c 3}
+                    nil)
+        _ (fact (grab-history)
+                => [{:op :replace!, :prop :a, :instance :obj, :coerced-new-value "1", :old-value 1, :new-value 1}
+                    {:op :replace!, :prop :b, :instance :obj, :coerced-new-value "2", :old-value 2, :new-value 2}
+                    {:op :replace!, :prop :c, :instance :obj, :coerced-new-value "3", :old-value 3, :new-value 3}])
+        _ (fact (component/instance component)
+                => :obj)
+
+        ;; delete
+        component (lifecycle/delete
+                    lifecycle
+                    component
+                    nil)
+        _ (fact (grab-history)
+                => [;;FIXME
+                    ;{:op :set-prop, :prop :a, :v nil}
+                    ;{:op :set-prop, :prop :b, :v nil}
+                    ;{:op :set-prop, :prop :c, :v nil}
+                    ])
```

Comment:
This is intentional btw. Cljfx is not a generic engine for transforming data to side effects on mutable objects — it's tailored for JavaFX object graph specifically. The purpose of the delete operation is to do a minimum amount of necessary cleanup when we don't need something anymore. There is no point in resetting properties of a TextField to initial values when you remove it from the display graph because you no longer need it. delete is mostly used for stuff like closing open windows when those are deleted.

Reply:
Yes, makes complete sense to me. Thanks for explaining.